### PR TITLE
feat(libsinsp): plugin dynamic reconfiguration

### PR DESCRIPTION
**Pull request by mrgian**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
feat(libsinsp): plugin dynamic reconfiguration
```

### CODE REVIEW

**Review by FedeDP — APPROVED** [review]
/approve

File: userspace/libsinsp/plugin.cpp [hunk 1]
```diff
@@ -712,6 +713,18 @@ void sinsp_plugin::validate_init_config_json_schema(std::string& config, std::st
...
 
+void sinsp_plugin::set_config(const std::string& config)
+{
+	std::string conf = config;
+	validate_config(conf);
+
+	if(m_state && m_handle->api.set_config)
+	{
+		m_handle->api.set_config(m_state, conf.c_str());
+		m_state = NULL;
```
Comment (jasondellaluce) [inline userspace/libsinsp/plugin.cpp:724]:
```suggestion
```
Why is the state set to null?

File: userspace/libsinsp/plugin.cpp [hunk 2]
```diff
@@ -712,6 +713,18 @@ void sinsp_plugin::validate_init_config_json_schema(std::string& config, std::st
 	}
 }
 
+void sinsp_plugin::set_config(const std::string& config)
+{
+	std::string conf = config;
```
Comment (jasondellaluce) [inline userspace/libsinsp/plugin.cpp:718]:
```suggestion
	if (!m_inited)
	{
		throw sinsp_exception(std::string(s_not_init_err) + ": " + m_name);
	}
	std::string conf = config;
```
I would enforce the plugin to be already initialized as a requirement.

File: userspace/libsinsp/plugin.cpp [hunk 3]
```diff
@@ -712,6 +713,18 @@ void sinsp_plugin::validate_init_config_json_schema(std::string& config, std::st
...
+	validate_config(conf);
+
+	if(m_state && m_handle->api.set_config)
```
Comment (jasondellaluce) [inline userspace/libsinsp/plugin.cpp:721]:
```suggestion
	if(m_handle->api.set_config)
```
I don't think that having a non-NULL state should be a requirement. The API currently allows NULL states in theory, so let's make the plugin decide whether to accept it or not.

File: userspace/plugin/plugin_api.h [hunk 4]
```diff
@@ -939,6 +939,14 @@ typedef struct
...
 		ss_plugin_rc (*set_async_event_handler)(ss_plugin_t* s, ss_plugin_owner_t* owner, const ss_plugin_async_event_handler_t handler);
 	};
+
+	// Sets a new plugin configuration when provided by the framework.
+	// Required: no
+	// Arguments:
+	// - s: the plugin state, returned by init(). Can be NULL.
+	// - config: the new configuration, as an opaque string.
+	//
+	void (*set_config)(ss_plugin_t* s, const char* config);
```
Comment (jasondellaluce) [inline userspace/plugin/plugin_api.h:949]:
```suggestion
	ss_plugin_rc (*set_config)(ss_plugin_t* s, const char* config);
```
I think we need a way to report errors someway, take a look at how we did it in other API symbols. The plugin should have the right to reject the config and provide context on the error with `get_last_error`.

Also (not strong on this point), wouldn't it be more appropriate to pass in a `ss_plugin_set_config_input*` or something, instead than the string itself? Over time, it happened in may occasions that new features required passing more parameters than expected, so I'm always cheering for the most future-proof solution.

**Review by jasondellaluce — CHANGES_REQUESTED** [review]
Thanks Gian! Left you some comments

File: userspace/libsinsp/plugin.cpp (hunk 1 above)
Comment (mrgian) [reply userspace/libsinsp/plugin.cpp:724]: Oops! A leftover from some prior experiments

File: userspace/libsinsp/plugin.cpp [hunk 5]
```diff
@@ -712,6 +713,28 @@ void sinsp_plugin::validate_init_config_json_schema(std::string& config, std::st
...
+	ss_plugin_rc rc = SS_PLUGIN_FAILURE;
+	ss_plugin_set_config_input input;
+	input.config = conf.c_str();
+
+	if(m_handle->api.set_config)
+	{
+		rc = m_handle->api.set_config(m_state, &input);
+	}
+
+	return rc == SS_PLUGIN_SUCCESS;
```
Comment (jasondellaluce) [inline userspace/libsinsp/plugin.cpp:735]:
```suggestion
	ss_plugin_set_config_input input;
	input.config = conf.c_str();

	if(!m_handle->api.set_config)
	{
		return false;
	}

	return m_handle->api.set_config(m_state, &input) == SS_PLUGIN_SUCCESS;
```

If there is no API symbol, I guess we should let users know that the config has not been set.

**Review by jasondellaluce — COMMENTED** [review]
Last suggestion

**Review by jasondellaluce — APPROVED** [review]
/approve

**Review by incertum — APPROVED** [review]
/approve

**Pull request merged**